arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -1078,104 +1078,138 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
     _config.barrier.extend(_config.connection);
   }
 
+  // We'll do one quick attempt at getting the keys on the leader.
+  // We might receive the keys or a count of the keys. In the first case we continue
+  // with the sync. If we get a coument count, we'll estimate a pessimistic wait and
+  // repeat the call without a quick option.
   std::string const baseUrl = replutils::ReplicationUrl + "/keys";
-  std::string url = baseUrl + "/keys" + "?collection=" + urlEncode(leaderColl) +
+  std::string url = baseUrl + "?collection=" + urlEncode(leaderColl) +
                     "&to=" + std::to_string(maxTick) +
                     "&serverId=" + _state.localServerIdString +
                     "&batchId=" + std::to_string(_config.batch.id);
 
-  std::string msg =
-      "fetching collection keys for collection '" + coll->name() + "' from " + url;
-  _config.progress.set(msg);
+  VPackBuilder builder;
+  VPackSlice slice;
+  auto maxWaitTime = _config.applier._initialSyncMaxWaitTime;
 
-  // send an initial async request to collect the collection keys on the other
-  // side
-  // sending this request in a blocking fashion may require very long to
-  // complete,
-  // so we're sending the x-arango-async header here
-  auto headers = replutils::createHeaders();
-  headers[StaticStrings::Async] = "store";
+  auto keysCall = [&](bool quick) {
 
-  std::unique_ptr<httpclient::SimpleHttpResult> response;
-  _config.connection.lease([&](httpclient::SimpleHttpClient* client) {
-    response.reset(client->retryRequest(rest::RequestType::POST, url, nullptr, 0, headers));
-  });
+    std::string msg =
+      "fetching collection keys for collection '" + coll->name() + "' from " + url;
+    _config.progress.set(msg);
 
-  if (replutils::hasFailed(response.get())) {
-    return replutils::buildHttpError(response.get(), url, _config.connection);
-  }
+    // send an initial async request to collect the collection keys on the other
+    // side
+    // sending this request in a blocking fashion may require very long to
+    // complete,
+    // so we're sending the x-arango-async header here
+    auto headers = replutils::createHeaders();
+    headers[StaticStrings::Async] = "store";
 
-  bool found = false;
-  std::string jobId = response->getHeaderField(StaticStrings::AsyncId, found);
+    LOG_DEVEL << ((quick) ? url + "&quick=true" : url);
```

Comment:
Either remove the LOG_DEVEL or change it into a debug log.

Reply:
Surely